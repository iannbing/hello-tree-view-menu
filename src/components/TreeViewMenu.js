import React, { Component } from 'react';
import { Button } from 'reactstrap';

class TreeViewMenu extends Component {
  static defaultProps = { data: null, path: '' };

  componentDidMount() {
    const { data } = this.props;

    console.log(data);
  }

  render() {
    const { path } = this.props;

    return (
      <div>
        <div>{path}</div>
        <Button color="primary">Click</Button>
      </div>
    );
  }
}

export default TreeViewMenu;
