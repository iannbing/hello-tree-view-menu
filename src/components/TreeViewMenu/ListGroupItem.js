import React from 'react';
import styled from 'react-emotion';
import { ListGroupItem as DefaultListGroupItem } from 'reactstrap';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1;

const ToggleIcon = styled('div')({
  position: 'absolute',
  left: 20,
  width: 10,
  height: 10
});

const ListGroupItemWithPadding = styled(DefaultListGroupItem)(
  ({ level = 0 }) => ({
    paddingLeft: `${DEFAULT_PADDING + level * LEVEL_SPACE}rem`
  })
);

const ListGroupItem = ({
  hasSubItems = false,
  isOpen = false,
  level = 0,
  children,
  ...props
}) => (
  <>
    <ListGroupItemWithPadding level={hasSubItems ? level : 0} {...props}>
      {hasSubItems && <ToggleIcon>{isOpen ? '-' : '+'}</ToggleIcon>}
      {children}
    </ListGroupItemWithPadding>
  </>
);

export default ListGroupItem;
