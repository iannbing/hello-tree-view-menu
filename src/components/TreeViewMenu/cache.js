const memory = {};

const cache = {
  get: key => memory[key],
  set: (key, value) => {
    memory[key] = value;
  }
};

export default cache;
