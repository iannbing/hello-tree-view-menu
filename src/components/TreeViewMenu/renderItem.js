import React from 'react';
import styled from 'react-emotion';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1.25;
const ICON_SIZE = 1;
const DefaultToggleIcon = ({ on }) => <div>{on ? '-' : '+'}</div>;

const ToggleIconContainer = styled('div')(({ level = 0 }) => ({
  position: 'absolute',
  left: `${ICON_SIZE + level * LEVEL_SPACE}rem`,
  width: `${ICON_SIZE}rem`,
  height: `${ICON_SIZE}rem`
}));

const ListItemContainer = styled('li')(({ level = 0 }) => ({
  paddingLeft: `${DEFAULT_PADDING + ICON_SIZE + level * LEVEL_SPACE}rem`,
  cursor: 'pointer'
}));

export const ListGroup = styled('ul')({
  listStyleType: 'none',
  paddingLeft: 0
});

const Input = styled('input')({
  margin: '.5em',
  paddingLeft: '.4em'
});

export const renderSearch = onSearch => (
  <Input placeholder="Type and search" onChange={onSearch} />
);

const renderItem = ({
  hasSubItems = false,
  isOpen = false,
  level = 0,
  onClick,
  active,
  key,
  toggleIcon = DefaultToggleIcon,
  label,
  ...props
}) => {
  const ToggleIcon = toggleIcon;

  return (
    <ListItemContainer
      level={level}
      onClick={onClick}
      active={active}
      key={key}
      {...props}
    >
      {hasSubItems && (
        <ToggleIconContainer level={level}>
          <ToggleIcon on={isOpen} />
        </ToggleIconContainer>
      )}
      {label}
    </ListItemContainer>
  );
};

export default renderItem;
