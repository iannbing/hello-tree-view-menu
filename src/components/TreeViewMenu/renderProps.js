import React from 'react';
import styled from 'react-emotion';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1.25;
const ICON_SIZE = 1;
const ToggleIcon = ({ on }) => <div>{on ? '-' : '+'}</div>;

const ToggleIconContainer = styled('div')(({ level = 0 }) => ({
  position: 'absolute',
  left: `${ICON_SIZE + level * LEVEL_SPACE}rem`,
  width: `${ICON_SIZE}rem`,
  height: `${ICON_SIZE}rem`
}));

const ListItemContainer = styled('li')(({ level = 0, active }) => ({
  paddingLeft: `${DEFAULT_PADDING + ICON_SIZE + level * LEVEL_SPACE}rem`,
  cursor: 'pointer',
  color: active ? 'white' : '#333',
  background: active ? '#179ed3' : 'none'
}));

const ListGroup = styled('ul')({
  listStyleType: 'none',
  paddingLeft: 0
});

export const renderGroup = items => <ListGroup>{items}</ListGroup>;

const Input = styled('input')({
  margin: '.5em',
  paddingLeft: '.4em'
});

export const renderSearch = onSearch => (
  <Input placeholder="Type and search" onChange={onSearch} />
);

export const renderItem = ({
  hasSubItems = false,
  isOpen = false,
  level = 0,
  onClick,
  active,
  key,
  label,
  ...props
}) => (
  <ListItemContainer
    level={level}
    onClick={onClick}
    active={active}
    key={key}
    {...props}
  >
    {hasSubItems && (
      <ToggleIconContainer level={level}>
        <ToggleIcon on={isOpen} />
      </ToggleIconContainer>
    )}
    {label}
  </ListItemContainer>
);
