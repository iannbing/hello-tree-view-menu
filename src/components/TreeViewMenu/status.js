const memoryStatus = {
  isInitializing: true
};

const status = {
  get: key => memoryStatus[key],
  set: (key, value) => {
    memoryStatus[key] = value;
  }
};

export default status;
