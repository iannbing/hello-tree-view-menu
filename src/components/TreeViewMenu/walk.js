const walk = ({ data, ...props }) =>
  Object.entries(data)
    .sort((a, b) => a[1].index - b[1].index)
    .reduce(
      (all, [nodeName, node]) => [
        ...all,
        ...(node.key ? generateBranch(node, nodeName, props) : []),
      ],
      [],
    );

const generateBranch = (node, nodeName, props) => {
  const { parent = '', level = 0, openNodes, searchTerm } = props;
  const { nodes, label } = node;
  const nodePath = [parent, nodeName].filter(x => x).join('/');
  const isOpen = !!nodes && (openNodes.includes(nodePath) || !!searchTerm);
  const isVisible =
    !searchTerm ||
    label.toLowerCase().includes(searchTerm.trim().toLowerCase());

  const currentItem = isVisible && {
    isOpen,
    nodePath,
    ...props,
    ...node,
  };
  const nextLevelItems = isOpen
    ? walk({ data: nodes, ...props, parent: nodePath, level: level + 1 })
    : [];

  return [currentItem, ...nextLevelItems].filter(x => x);
};

export default walk;
