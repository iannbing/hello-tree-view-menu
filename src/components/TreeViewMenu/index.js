import React from 'react';
import { ListGroup } from 'reactstrap';

import ListGroupItem from './ListGroupItem';

const defaultOnClick = () => console.warn('no behavior defined'); // eslint-disable-line no-console

class TreeViewMenu extends React.Component {
  static defaultProps = {
    data: null,
    activeKey: '',
    searchTerm: ''
  };

  state = { openNodes: [] };

  toggleNode = node => {
    const { openNodes } = this.state;
    if (openNodes.includes(node)) {
      this.setState({
        openNodes: openNodes.filter(openNode => openNode !== node)
      });
    } else {
      this.setState({ openNodes: [...openNodes, node] });
    }
  };

  getOnClickFunction = ({ onClick, node }) => () => {
    const onClickFn = onClick || defaultOnClick;
    onClickFn();
    this.toggleNode(node);
  };

  walk = ({ data, parent = '', level = 0 }) => {
    const { activeKey, searchTerm } = this.props;
    const { openNodes } = this.state;

    return Object.entries(data)
      .sort((a, b) => a[1].index - b[1].index)
      .reduce((all, [nodeName, node]) => {
        const { label, onClick, nodes, key } = node;
        const hasSubItems = nodes !== undefined && nodes !== null;
        if (!key) return all;
        const isActive = activeKey === key;
        const currentNode = [parent, nodeName].filter(x => x).join('/');
        const isOpen = openNodes.includes(currentNode) || searchTerm;
        const onClickFunction = this.getOnClickFunction({
          onClick,
          node: currentNode
        });
        const isMatching = searchTerm && label.includes(searchTerm);
        const currentItem = (!searchTerm || isMatching) && (
          <ListGroupItem
            hasSubItems={hasSubItems}
            isOpen={isOpen}
            level={level}
            onClick={onClickFunction}
            active={isActive}
            key={currentNode}
          >
            {label}
          </ListGroupItem>
        );
        const nextLevelItems =
          nodes &&
          isOpen &&
          this.walk({
            data: nodes,
            parent: currentNode,
            level: level + 1
          });
        return [...all, currentItem, nextLevelItems].filter(x => x);
      }, []);
  };

  render() {
    const { data, activeKey } = this.props;
    const { openNodes } = this.state;
    return (
      data && <ListGroup>{this.walk({ data, activeKey, openNodes })}</ListGroup>
    );
  }
}

export default TreeViewMenu;
