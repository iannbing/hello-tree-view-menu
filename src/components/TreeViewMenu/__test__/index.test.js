import React from 'react';
import { shallow } from 'enzyme';

import TreeViewMenu from '../index';
import data from '../../../data/spaces.json';

describe('TreeViewMenu', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<TreeViewMenu data={data} path="" />);

    expect(wrapper).toMatchSnapshot();
  });
});
