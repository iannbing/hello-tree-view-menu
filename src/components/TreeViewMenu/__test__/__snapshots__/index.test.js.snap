// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeViewMenu should render correctly 1`] = `
<ListGroup
  tag="ul"
>
  <Styled(ListGroupItem)
    active={true}
    onClick={[Function]}
    tag="li"
  >
    Release Notes
  </Styled(ListGroupItem)>
  <Styled(ListGroupItem)
    level={1}
    tag="li"
  >
    Developer Portal
  </Styled(ListGroupItem)>
</ListGroup>
`;
